include/linux/kfence.h

```diff
@@ -68,14 +69,31 @@ static inline void kfence_observe_memcg_cache(struct kmem_cache *memcg_cache) {
 
 #ifdef CONFIG_KFENCE_NAIVE
 
+DECLARE_PER_CPU(int, kfence_sample_cnt);
+extern unsigned long kfence_sample_rate;
+
 void *kfence_alloc_with_size(struct kmem_cache *s, size_t size, gfp_t flags);
+static __always_inline void *kfence_alloc_with_addr(struct kmem_cache *s, gfp_t flags, unsigned long addr)
+{
+	int cnt = this_cpu_dec_return(kfence_sample_cnt);
+	size_t size;
+
+	if (likely(cnt > 0))
+		return NULL;
+	this_cpu_write(kfence_sample_cnt, kfence_sample_rate);
+	size = addr >> 48;
```

Comment:
This assume that 'addr' is 64bit, which I don't think we can do.

You could calculate this as 'BITS_PER_LONG - 16'.

Reply:
Ack.